switools/swisignature.py

```diff
@@ -144,9 +160,123 @@ def signSwiHandler( args ):
+                  )
+   if ret != 0:
+      print( "Error: Cannot extract signature file %s from  %s" % ( sigFn, swi ) )
+      sys.exit( SWI_SIGN_RESULT.CANNOT_EXTRACT_SIGNATURE )
+
+def getSignatureFile( swi, signatureFile ):
+   serviceName = 'swi-signing-service'
+   serviceBinary = shutil.which( serviceName )
+   if not serviceBinary:
+      print( "Error: signing service '%s' not found" % serviceName )
+      sys.exit( SWI_SIGN_RESULT.ERROR_SIGNING_SERVER_FAILED )
+
+   sha256 = prepareSwi( swi=swi, outfile=None, forceSign=True )
+   print( "%s sha256: %s" % ( os.path.basename( swi ), sha256 ) )
+   ret = subprocess.check_call( [ 'swi-signing-service', sha256, signatureFile ] )
+   if ret != 0:
+      print( "Error: signing-server '%s' failed" % serviceBinary )
+      sys.exit( SWI_SIGN_RESULT.ERROR_SIGNING_SERVER_FAILED )
+
+def signSwiAll( swi, signingCertFile, rootCaFile, signatureFile=None, signingKeyFile=None ):
+   # Sub-images ("optimizations") are extracted to /tmp. The utility 'swadapt' is
+   # handling that extraction. swadapt is found inside the image itself and is a
+   # statically linked i386 binary.
+   workDir = "/tmp/optims-%d" % os.getpid()
+   os.mkdir( workDir )
+   # Make sure the image we got is a swi file
+   if ( not os.path.isfile( swi ) or
+        os.system( "set -e; image=$(readlink -f %s); cd %s;"
+                   "unzip -o -q $image version" % ( swi, workDir ) ) ):
+      print( "Error: '%s' does not look like an EOS image" % swi )
+      shutil.rmtree( workDir )
+      sys.exit( SWI_SIGN_RESULT.ERROR_NOT_A_SWI )
+
+   try:
+      optims = signaturelib.getOptimizations( swi, workDir )
+      if optims is None or len( optims ) == 1 or "DEFAULT" in optims:
+         # legacy case of a single rootfs image
+         # maybe need to use a remote signing service (new feature in v1.2)
+         if not signatureFile and not signingKeyFile:
+            signatureFile = "%s/sig" % workDir
+            getSignatureFile( swi, signatureFile ) # never returns in case of fail
+         if not signatureFile and not signingKeyFile:
+            print ( 'Error: without signing key we need a signature file: '
+                    'run "swi-signature prepare" first and have the digest it prints '
+                    'signed and passed to this command.' )
+            sys.exit( SWI_SIGN_RESULT.ERROR_NO_SIGNATURE_FILE_PROVIDED )
+         if not signatureFile and signingKeyFile:
+            sha256 = prepareSwi( swi=swi, outfile=None, forceSign=True )
+            print( "%s sha256: %s" % ( os.path.basename( swi ), sha256 ) )
+         # insert the signatureFile into the swi, or use the signing key to create the signatureFile first
+         return signSwi( swi, signingCertFile, rootCaFile,
+                         signatureFile=signatureFile, signingKeyFile=signingKeyFile )
+
+      print( "Optimizations in %s: %s" % ( swi, " ".join( optims ) ) )
+      signaturelib.extractSwadapt( swi, workDir )
```

Comment:
Earlier I suggested moving the error handling out of `signaturelib.extractSwadapt` and passing it a `zipfile.ZipFile` object, so this would become: 
```suggestion
      if not signaturelib.extractSwadapt( zf, workDir ):
         print( "Error: '%s' does not contain the 'swadapt' utility" % swi )
         sys.exit( SWI_SIGN_RESULT.ERROR_MISSING_SWADAPT )
```